DOXYGEN: Review of high prio files from graphics

b-gent opened the pull request:
Editing doxygen comments in:

- cursorman.h
- font.h
- managed_surface.h
- pixelformat.h
- surface.h

Still working on other files but pushing what I have now.

<!---
Thank you for contributing to ScummVM. Please read the following carefully before submitting your Pull Request.

Make sure your individual commits follow the guidelines found in the ScummVM Wiki: https://wiki.scummvm.org/index.php?title=Commit_Guidelines. If they're not please edit them before submitting the Pull Request.

Commits and Pull Requests should use the following template:

```
SUBSYSTEM: Short (50 chars or less) summary of changes

More detailed explanatory text, if necessary.  Wrap it to about 72
characters or so.  In some contexts, the first line is treated as the
subject of an email and the rest of the text as the body.  The blank
line separating the summary from the body is critical (unless you omit
the body entirely); tools like rebase can get confused if you run the
two together.

Write your commit message in the present tense: "Fix bug" and not "Fixed
bug."  This convention matches up with commit messages generated by
commands like git merge and git revert.

Further paragraphs come after blank lines.

- Bullet points are okay, too

- Typically a hyphen or asterisk is used for the bullet, preceded by a
 single space, with blank lines in between, but conventions vary here

- Use a hanging indent
```
--->

criezy commented on graphics/cursorman.h:
```diff
@@ -31,19 +31,28 @@
-	 * This fact might result in a non visible mouse cursor if the caller does
-	 * not call OSystem::updateScreen itself after a showMouse(true) call.
+	 * This might result in a non-visible mouse cursor if the caller does
+	 * not call OSystem::updateScreen after a showMouse(true) call.
```
Maybe rewrite that sentence to improve the flow.
I would for example suggest something like: `You need to call OSystem::updateScreen after a call to showMouse(true) to ensure that the mouse cursor becomes visible.`

criezy commented on graphics/font.h:
```diff
@@ -112,91 +126,103 @@ class Font {
 
-	// TODO: Add doxygen comments to this
+	/**
+	 * @todo Doc required!
```
All the `drawString` functions below, as their name suggest, draw the given `str` string to the given `dst` surface.
The arguments are the same ones as in the `getBoundingBox` above, with a few additional ones:
- `dst` The surface on which to draw the string.
- `str`  The string to draw.
- `x`    The x position where to start drawing.
- `y`    The y position where to start drawing.
- `w`   The width of the text area.
- `color` The colour with which to draw the string.
- `align` The text alignment. This can be used to center the string in the given area or to align it to the right.
- `deltax` Offset to the x starting position of the string.
- `useEllipsis` Use ellipsis if needed to fit the string in the area.

criezy reviewed (changes requested):
There are a few minor corrections to make, and a few undocumented functions to document.
But otherwise this looks quite good already.

b-gent commented on graphics/cursorman.h:
```diff
@@ -120,59 +136,68 @@ class CursorManager : public Common::Singleton<CursorManager> {
 private:
+	/**
+	* Generic class for implementing the singleton design pattern.
+	*/
```
Yes, and the description is taken from that file.

With the current config, doxygen shows friend classes in the output and I just want everything in the output to be documented.

![image](https://user-images.githubusercontent.com/29537761/100398893-e60c1300-3050-11eb-9c77-0d5352415662.png)

 I'm not trying to say it's declared here, just providing the description of what the class is. I could also replace with just a reference to the doc of this class - "See @ref (...)"

Another question is, should we show friend classes in the doc output. Maybe it's completely unnecessary.

criezy replied:
I think it is unnecessary to show it. This is an internal implementation details for the class. Developers using the CursorManager do not need to know that.

b-gent replied:
Ok, I will make sure to disable showing friend classes in an upcoming PR.

b-gent replied:
Commit added here now.

b-gent commented:
Hi @criezy  , here are today's updates:

1. Added managed_surface.h, pixelformat.h, and surface.h to the initial commit. Pixelformat gave me a lot of headache and I had to do a ton of reading to understand the whole concept, but I was still very hesitant when adding comments there. Please review.
2. Added a second commit that adds doxy groups to the other headers in the graphics folder. I only did it for the ones that contain some doxygen. The descriptions for the groups do not matter much right now, I will be reviewing these files later anyway, so unless they are very wrong, you do not need to review them.
3. I uploaded HTML output to here:
https://b-gent.github.io/2645/d8/d40/group__graphics__conversion.html
The navi looks like this right now, which I think is pretty neat:
![image](https://user-images.githubusercontent.com/29537761/100530333-f0651300-31f0-11eb-963b-ef7713233330.png)

criezy commented on graphics/pixelformat.h:
```diff
@@ -28,18 +28,27 @@
 struct ColorComponent {
 };
-
+/** @todo Doc required (0-bit element?) */
```
I am not quite sure why we have a specialisation for 0.  Maybe it is here for completeness or it might be needed if we have formats without alpha channel for example (although the code in `colorToARGB` seems to be checking for that explicitly to use 0xFF, which makes more sense than 0).

b-gent replied:
I coud use 'Return 0 for an empty color component.' ? Does that make sense?

criezy replied:
Yes, I think it does.

criezy commented on graphics/pixelformat.h:
```diff
@@ -111,7 +120,7 @@ struct ColorComponent<7> {
 	}
 };
-
+/** Return.... @todo not sure what that does if it's already 8-bit? */
```
It's just returning the given value (and making it doesn't use more than 8 bits).
I think it is mostly here so that 8 bits component do not have to be treated in a specific way and we can use the same code for all the cases.

criezy commented on graphics/pixelformat.h:
```diff
@@ -139,12 +148,14 @@ struct PixelFormat {
 	byte rLoss, gLoss, bLoss, aLoss; /**< Precision loss of each color component. */
 	byte rShift, gShift, bShift, aShift; /**< Binary left shift of each color component in the pixel value. */
 
+	/** Define a pixel format ... @todo Doc required */
```
Default constructor that creates a NULL pixel format.

criezy commented on graphics/pixelformat.h:
```diff
@@ -139,12 +148,14 @@ struct PixelFormat {
 		rShift = gShift = bShift = aShift = 0;
 	}
 
+	/** Define a pixel format ... @todo Is this RGB888? */
```
It is not necessarily RGB888. It depends on the value you pass to the constructor.
RGB888 would be `BytesPerPixel` = 3, `RBits` = `GBits` = `BBits` = 8, `ABits` = 0, `RShift` = 16, `GShift` = 8, `BShift` = 0, `AShift = 0

So basically `BytesPerPixel` is the total number of bytes used to define a colour, `RBits`, `GBits`, `BBits`, and `ABits` are the number of bits for each component, and `RShift`, `GShift`, `BShift`, and `AShift` is the amount of bit shift for the component in the colour value.

Some other examples:
- RGBA8888 is `BytesPerPixel` = 4, `RBits` = `GBits` = `BBits` = `ABits` = 8, `RShift` = 24, `GShift` = 16, `BShift` = 8, `AShift = 0
- ABGR8888 is `BytesPerPixel` = 4, `RBits` = `GBits` = `BBits` = `ABits` = 8, `RShift` = 0, `GShift` = 8, `BShift` = 16, `AShift = 24
- RGB565 is `BytesPerPixel` = 2, `RBits` = 5, `GBits` = 6, `BBits` = 5, `ABits` = 0, `RShift` = 11, `GShift` = 5, `BShift` = 0, `AShift = 0

criezy commented on graphics/pixelformat.h:
```diff
@@ -196,64 +212,96 @@ struct PixelFormat {
 	}
 
+	/**
+	 * Return the number of component bits from the specified bits per pixel (bpp) value.
```
Which specified `bpp` value? The function does not take any argument.
It returns the total number of bits for the pixel format.

criezy commented on graphics/surface.h:
```diff
@@ -134,223 +145,234 @@ struct Surface {
+	 * Move the surface horizontally or vertically.
+	 *
+	 * See @c surface.cpp for the implementation.
+	 * @todo More explanation required?
```
It would be more accurate to say that it moves the surface content horizontally and/or vertically by the given number of pixels. And I don't think you need more explanation than that.

criezy commented on graphics/managed_surface.h:
```diff
@@ -30,143 +30,159 @@
 	/**
-	 * Inner method for copying another surface into this one at a given destination position
+	 * Inner method for copying another surface into this one at a given destination position.
+	 * @todo there are a lot of parameters here, are they all fairly clear or should we document them?
```
I don't think you need to document them. As the description indicates, it is an internal function for the class. Code using a `ManagedSurface` will not use this function directly.

criezy commented on graphics/managed_surface.h:
```diff
@@ -30,143 +30,159 @@
+	 * directly to it, since it would bypass dirty rect handling.
 	 */
 	operator const Surface &() const { return _innerSurface; }
+	/** @todo not sure what that rawSurface is */
```
The `ManagedSurface` class is a wrapper around a `Surface` object. The `rawSurface` is simply this `Surface` object that is being wrapped.

b-gent replied:
I copied the comment from the method above.

Question - is this technically an ``@overload`` if this method has a different name but does the same thing as the other method?

criezy replied:
Technically an overloaded functions are functions with the same name but different arguments. So no, it is not an overload.
Also, as you can see, the method above doesn't actually have a name. This is something called  a `conversion operator` in C++.

criezy reviewed (changes requested):
There are only a few changes needed.
Overall it looks close to a state in which it can be merged.

b-gent commented:
Comments are fixed now, branch is rebased. I left two small questions in the unresolved comments, please anwer whenever you have time @criezy

criezy reviewed (approved):
It looks good to merge now. Thank you!
